packages/react-components/react-card/src/components/CardHeader/useCardHeader.ts

```diff
@@ -21,15 +21,25 @@ export const useCardHeader_unstable = (props: CardHeaderProps, ref: React.Ref<HT
   } = useCardContext_unstable();
   const headerRef = React.useRef<HTMLDivElement>(null);
 
+  const hasChildId = React.useRef(false);
   const generatedId = useId(cardHeaderClassNames.header, referenceId);
 
   React.useEffect(() => {
-    if (header && headerRef.current) {
-      const { id } = headerRef.current;
+    const refId = !hasChildId.current && headerRef.current?.id;
```

Comment:
hmm after taking a bit more "deep dive look" we can make the use effect callback shorter and without type assertions ( which are usually a red flag ).


```ts
React.useEffect(() => {
    // be explicit . using ecma shortcircuit operator `&&`  creates misleading data types
    const refId = !hasChildId.current ? headerRef.current?.id : undefined;
    const childWithId = getChildWithId(header);

    hasChildId.current = Boolean(childWithId);

    setReferenceId(getReferenceId(refId, childWithId, generatedId));
  }, [generatedId, header, setReferenceId]);
``` 

now those helper functions can be hoisted to root scope within this module 

```ts
function getChildWithId(header: CardHeaderProps['header']) {
  // if we use predicate TS will properly narrow down `find` call
  function isReactElementWithIdProp(element: React.ReactNode): element is React.ReactElement {
    return React.isValidElement(element) && Boolean(element.props.id);
  }

  return React.Children.toArray(header).find(isReactElementWithIdProp);
}

function getReferenceId(
  refId: string | undefined,
  childWithId: React.ReactElement | undefined,
  generatedId: string,
): string {
  if (refId) {
    return refId;
  }
  return childWithId?.props.id ? childWithId.props.id : generatedId;
}
```

Reply:
Nice little trick with the `element is React.ReactElement`. I like that!